src/plugins/segmentio/__tests__/batched-dispatcher.test.ts

```diff
@@ -49,10 +49,13 @@ describe('Batching', () => {
   beforeEach(() => {
     jest.resetAllMocks()
     jest.restoreAllMocks()
+    jest.useFakeTimers()
   })
 
   afterEach(() => {
-    window.close()
+    // clear any pending sendBatch calls
+    jest.runAllTimers()
```

Comment:
Does jest.clearAllTimers() work here instead?

Reply:
I tried that, but for these tests we need the timers to actually run to completion so that the batched dispatcher flushes all its events.

If I just clear the timeouts instead what ends up happening is later on when we dispatch a `beforeunload` event it triggers the batched dispatchers we created in earlier tests to flush their events and this state leaks into other tests.

`window.close()` was triggering the `beforeunload` before so it wasn't a problem but since we can't use that anymore (unless each test is in its own file or making updates to jest's teardown) running all timers ensures the buffered events are flushed before starting the next test.

Reply:
For a module we're there's some impure state, I think it's best to have the  import of the module as part of the beforeEach (and then reset in the afterEach), so each test case is isolated with a "fresh" import.

Reply:
I tried some simple refactoring but this looks like it'll be more involved. I think ultimately out of scope for just upgrading jest :smile: